File: usr-share/controlpanel/repository.js
/* begin license *
 *
 * "Metastreams Harvester" is a fork of Meresco Harvester that demonstrates
 * the translation of traditional metadata into modern events streams.
 *
 * Copyright (C) 2021 Seecr (Seek You Too B.V.) https://seecr.nl
 *
 * This file is part of "Metastreams Harvester"
 *
 * "Metastreams Harvester" is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * "Metastreams Harvester" is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with "Metastreams Harvester"; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * end license */

function init_cardRepositoryAttributes() {
    var _frm = $("#FrmRepositoryAttributes");
    var _btn = $("#BtnRepositoryAttributes");
    _btn
        .prop("disabled", true)
        .unbind("click")
        .click(function(e) {
            e.preventDefault();
            $.post("/action/updateRepositoryAttributes", _frm.serialize())
                .done(function(data) {
                    if (data['success'] == true) {
                        form_resetBordersAndDisabled(_frm, _btn);
                    } else {
                        msg_Error(placeholder=$("#placeholder_FrmRepositoryAttributes"), identifier=undefined, text=data['message']);
                    }
                })
        });
    form_setBordersAndDisabled(_frm, _btn);
    form_resetBordersAndDisabled(_frm, _btn);
}

function init_cardClosingHours() {
    var _frm = $("#FrmCreateClosingHours");
    var _btn = $("#BtnCreateClosingHours");

    function _load_table_closinghours(domainId, repositoryId) {
        var _placeholder = $("#placeholder_table_closinghours");
        $.get("/repository/table/closingHours", $.param({domainId: domainId, identifier: repositoryId}))
            .done(function(page) {
                _placeholder
                    .empty()
                    .append(page);
                init_table_closinghours();
            });
    }


    function init_table_closinghours() {
        var _placeholder = $("#placeholder_table_closinghours");
        _placeholder.find("button.deletable.seecr-btn")
            .unbind("click")
            .click(function(e) {
                var _btn = $(this);
                var _domainId = _btn.data('domainid');
                var _repositoryId = _btn.data('repositoryid'); 
                e.preventDefault();
                $.post("/action/deleteRepositoryClosingHours", $.param({
                        repositoryId: _repositoryId,
                        domainId: _domainId,
                        closingHour: _btn.data('closinghour')
                    }))
                    .done(function(data) {
                        if (data['success'] == true) {
                            _load_table_closinghours(_domainId, _repositoryId);
                        } else {
                            msg_Error(placeholder=$("#placeholder_FrmCreateClosingHours"), identifier=undefined, text=data['message']);
                        }
                    })

            })
    }
    init_table_closinghours();

    _btn
        .prop("disabled", true)
        .unbind("click")
        .click(function(e) {
            e.preventDefault();
            $.post("/action/addRepositoryClosingHours", _frm.serialize())
                .done(function(data) {
                    if (data['success'] == true) {
                        form_resetBordersAndDisabled(_frm, _btn, true);
                        _load_table_closinghours(_frm.data("domainid"), _frm.data("repositoryid"));
                    } else {
                        msg_Error(placeholder=$("#placeholder_FrmCreateClosingHours"), identifier=undefined, text=data['message']);
                    }
                })
        });
    form_setBordersAndDisabled(_frm, _btn);
    form_resetBordersAndDisabled(_frm, _btn);
}

function init_cardRepositoryActions() {
    form_init(
        $("#FrmRepositoryActions"), "/action/updateRepositoryActionAttributes",
        $("#BtnRepositoryActions"), $("#placeholder_FrmRepositoryActions"));
}

function init_cardRepositoryFieldDefinitions() {
    form_init(
        $("#FrmFieldDefinition"), "/action/updateRepositoryFieldDefinitions",
        $("#BtnFieldDefinition"), $("#placeholder_FrmFieldDefinition"));
}

function init_cardHeaders() {
    let _frm = $("#FrmAddHeader");
    let _btn = $("#BtnAddHeader");

    function _load_table_headers(domainId, repositoryId) {
        var _placeholder = $("#placeholder_headers_list");
        $.get("/repository/table/headers", $.param({domainId: domainId, identifier: repositoryId}))
            .done(function(page) {
                _placeholder
                    .empty()
                    .append(page);
                init_table_headers();
            });
    }
    function init_table_headers() {
        var _placeholder = $("#placeholder_headers_list");
        _placeholder.find("button.deletable.seecr-btn")
            .unbind("click")
            .click(function(e) {
                e.preventDefault();
                var _btn = $(this);
                let _frm = _btn.closest("form");
                $.post("/action/remove_header", _frm.serialize())
                    .done(function(data) {
                        if (data['success'] == true) {
                            _load_table_headers(data['domainId'], data['repositoryId']);
                        } else {
                            msg_Error(placeholder=$("#placeholder_headers_add"), identifier=undefined, text=data['message']);
                        }
                    })
            })
    }

    _btn
        .unbind("click")
        .click(function(e) {
            e.preventDefault();
            $.post("/action/add_header", _frm.serialize())
                .done(function(data) {
                    if (data['success'] == true) {
                        _load_table_headers(data['domainId'], data['repositoryId']);
                        form_resetBordersAndDisabled(_frm, _btn, true);
                    } else {
                        msg_Error(placeholder=$("#placeholder_headers_add"), identifier=undefined, text=data['message']);
                    }
                })
        })
    init_table_headers();
    form_setBordersAndDisabled(_frm, _btn);
    form_resetBordersAndDisabled(_frm, _btn);
}

$(document).ready(function() {
    init_cardRepositoryAttributes();
    init_cardRepositoryActions();
    init_cardRepositoryFieldDefinitions();
    init_cardClosingHours();
    init_cardHeaders();


    init_button_status(); // from common.js

    $("button.helpText").each(function(index) {
        var _btn = $(this);
        _btn
            .unbind("click")
            .click(function(e) {
                e.preventDefault();
                var _modal = $("#modal");
                _modal.find("#placeholder_modal-title").html(_btn.data('caption'));
                $.get("/page2/modal/helpText", $.param({"text": _btn.data("helptext")}))
                    .done(function(data) {
                        _modal.find("#placeholder_modal-body")
                            .empty()
                            .append(data)
                        _modal.find("button")
                            .unbind("click")
                            .click(function(e) {
                                e.preventDefault();
                                _modal.modal("hide");
                            });
                    });
                _modal.modal("show");
            });
    })

})

File: meresco/harvester/repository.py
## begin license ##
#
# "Metastreams Harvester" is a fork of Meresco Harvester that demonstrates
# the translation of traditional metadata into modern events streams.
#
# Copyright (C) 2006-2007 SURFnet B.V. http://www.surfnet.nl
# Copyright (C) 2007-2008 SURF Foundation. http://www.surf.nl
# Copyright (C) 2007-2011 Seek You Too (CQ2) http://www.cq2.nl
# Copyright (C) 2007-2009 Stichting Kennisnet Ict op school. http://www.kennisnetictopschool.nl
# Copyright (C) 2009 Tilburg University http://www.uvt.nl
# Copyright (C) 2010-2011, 2015, 2020-2021 Stichting Kennisnet https://www.kennisnet.nl
# Copyright (C) 2015, 2017, 2020-2022 Seecr (Seek You Too B.V.) https://seecr.nl
# Copyright (C) 2020-2021 Data Archiving and Network Services https://dans.knaw.nl
# Copyright (C) 2020-2021 SURF https://www.surf.nl
# Copyright (C) 2020-2021 The Netherlands Institute for Sound and Vision https://beeldengeluid.nl
#
# This file is part of "Metastreams Harvester"
#
# "Metastreams Harvester" is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# "Metastreams Harvester" is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with "Metastreams Harvester"; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
## end license ##

from sys import exc_info
from traceback import format_exception
from time import localtime

from .action import Action, State
from .eventlogger import NilEventLogger
from .oairequest import OAIError, OaiRequest
from .saharaobject import SaharaObject
from .timeslot import Timeslot
from .virtualuploader import UploaderFactory

from gustos.common.units import COUNT
from functools import reduce

nillogger = NilEventLogger()


class Repository(SaharaObject):
    def __init__(self, domainId, repositoryId, oaiRequestClass=None):
        SaharaObject.__init__(self, [
            'repositoryGroupId', 'baseurl', 'set',
            'collection', 'metadataPrefix', 'use',
            'targetId', 'mappingId', 'action',
            'userAgent', 'authorizationKey',
            'complete', 'maximumIgnore', 'continuous'], ['shopclosed'])
        self.domainId = domainId
        self.id = repositoryId
        self.mockUploader = None
        self.uploadfulltext = True
        self._oaiRequestClass = oaiRequestClass or OaiRequest

    def closedSlots(self):
        if not hasattr(self, '_closedslots'):
            if self.shopclosed:
                self._closedslots = [Timeslot(txt) for txt in self.shopclosed]
            else:
                self._closedslots = []
        return self._closedslots

    def shopClosed(self, dateTuple = localtime()[:5]):
        return reduce(lambda lhs, rhs: lhs or rhs, [x.areWeWithinTimeslot( dateTuple) for x in self.closedSlots()], False)

    def target(self):
        return self._proxy.getTargetObject(domainId=self.domainId, identifier=self.targetId)

    def mapping(self):
        return self._proxy.getMappingObject(domainId=self.domainId, identifier=self.mappingId)

    def maxIgnore(self):
        return int(self.maximumIgnore) if self.maximumIgnore else 0

    def createUploader(self, logger):
        if self.mockUploader:
            return self.mockUploader
        return UploaderFactory().createUploader(self.target(), logger, self.collection)

    def oairequest(self):
        return self._oaiRequestClass(self.baseurl, userAgent=self.userAgent or None, authorizationKey=self.authorizationKey or None)

    def _createAction(self, repoState, generalHarvestLog):
        return Action.create(self, repoState, generalHarvestLog=generalHarvestLog)

    def do(self, stateDir, logDir, generalHarvestLog=nillogger, gustosClient=None):
        gustosReport = _prepareGustosReport(gustosClient, self.domainId, self.repositoryGroupId, self.id)
        repoState = None
        try:
            if not (stateDir or logDir):
                raise RepositoryException('Missing stateDir and/or logDir')
            repoState = State(stateDir, logDir, self.id)
            action = self._createAction(repoState, generalHarvestLog=generalHarvestLog)
            if action.info():
                generalHarvestLog.logLine('START',action.info(), id=self.id)
            actionIsDone, message, hasResumptionToken = action.do()
            if actionIsDone:
                self.action = None
                self._proxy.repositoryActionDone(self.domainId, self.id)
            if message:
                generalHarvestLog.logLine('END', message, id = self.id)
            completeHarvest = hasResumptionToken and self.complete == True
            if completeHarvest:
                generalHarvestLog.logInfo('Repository will be completed in one attempt', id=self.id)
            gustosReport(nrOfErrors=0)
            return message, completeHarvest
        except OAIError as e:
            gustosReport(nrOfErrors=1)
            errorMessage = _errorMessage()
            generalHarvestLog.logError(errorMessage, id=self.id)
            if e.errorCode() == 'badResumptionToken':
                action.resetState()
                return errorMessage, self.complete == True
            return errorMessage, False
        except:
            gustosReport(nrOfErrors=1)
            errorMessage = _errorMessage()
            generalHarvestLog.logError(errorMessage, id=self.id)
            return errorMessage, False
        finally:
            repoState is None or repoState.close()

def _prepareGustosReport(client, domain_id, repo_group_id, repo_id):
    if client is None:
        return lambda *_, **__: None
    def gustosReport(nrOfErrors):
        client.report(values={f'Harvester ({domain_id})': { f'{repo_group_id}:{repo_id}': { "errors": { COUNT: nrOfErrors}}}})
    return gustosReport

class RepositoryException(Exception):
    pass


def _errorMessage():
    xtype, xval, xtb = exc_info()
    return '|'.join(line.strip() for line in format_exception(xtype, xval, xtb))
